### PR TITLE
[HGCAL] Deprecate MultiCluster

### PR DESCRIPTION
**Pull request by lecriste**
#### PR description:

This PR replaces the usage of HGCal MultiCluster with Trackster.
No changes are expected, apart from
- naming shift ("MultiCluster" -> "Trackster") in DQM output.
- O(10%) difference in the MEs using the MultiCluster/Trackster energy (which may change of 0.01% for numerical precision) due to overflow migration; I verified that the unbinned histogram mean is identical. 

#### PR validation:

`runTheMatrix -l limited`

### CODE REVIEW

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
Ciao @lecriste thanks for this.
Would you mind taking care also of [this](https://github.com/cms-sw/cmssw/blob/master/RecoHGCal/Configuration/python/RecoHGCal_EventContent_cff.py#L6-L14)?

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
Since we have now 0 information in `AOD`, maybe we could just save the `_ticlTrackstersMerge_` there?
I'm not sure if it's for this PR or for the next one.
I'm not sure who is consuming `AOD` nowadays for `HGCAL` purposes. Maybe we could ask for that first, and, in case, add the suggested collection later on.
Maybe the reco convener knows a little more about `AOD` usage?

File: Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py [hunk 1]
```diff
@@ -5,28 +5,29 @@
...
-                   "ticlMultiClustersFromSimTracksters"]
+trackstersIters = ['ticlTracksters'+iteration for iteration in ticlIterLabelsMerge]
+trackstersIters.extend(["ticlSimTracksters"])
 
 simClustersGeneralLabel = 'simClusters'
 layerClustersGeneralLabel = 'hgcalLayerClusters'
-multiclustersGeneralLabel = 'hgcalMultiClusters'
-trackstersGeneralLabel = 'allTiclMultiClusters'
+trackstersGeneralLabel = 'hgcalTracksters'
+allTrackstersLabel = 'allTiclTracksters'
```
Comment (felicepantaleo) [inline Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py:23]: can the names and the collection options `simClusters`, `hgcalTracksters`, `allTiclTracksters` made more clear and explicit. 
It's not clear at all what collection should people use to validate tracksters..

**Review by felicepantaleo — CHANGES_REQUESTED** [review]

File: RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalTrackster.cc [hunk 2]
```diff
@@ -1,42 +1,42 @@
...
+    // Skip empty tracksters
+    if (tst.vertices().empty()) {
       continue;
     }
     // Filter using trackster PID
     if (filterByTracksterPID_) {
       float probTotal = 0.0f;
       for (int cat : filter_on_categories_) {
-        probTotal += tracksters[iMultiClus].id_probabilities(cat);
+        probTotal += tracksters[iTrackster].id_probabilities(cat);
```
Comment (rovere) [inline RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalTrackster.cc:39]: Why not using `tst` directly? Possibly you could even drop `iTrackster`.

File: RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalTrackster.cc [hunk 3]
```diff
@@ -47,17 +47,11 @@ void PFClusterFromHGCalMultiCluster::buildClusters(const edm::Handle<reco::PFRec
...
-
-    // Use the H&F of the clusters inside the multicluster if the latter's H&F are not stored
-    if (hitsAndFractions.empty()) {
-      for (const auto& cl : mcl) {
-        const auto& hAndF_temp = cl->hitsAndFractions();
-        hitsAndFractions.insert(hitsAndFractions.end(), hAndF_temp.begin(), hAndF_temp.end());
-      }
+    for (const auto lcId : tst.vertices()) {
+      const auto& hAndF_temp = clusters[lcId].hitsAndFractions();
+      hitsAndFractions.insert(hitsAndFractions.end(), hAndF_temp.begin(), hAndF_temp.end());
```
Comment (rovere) [inline RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalTrackster.cc:54]: Didn't you forget to apply the `multiplicity` here? I believe this is a bug.
See [here](https://github.com/cms-sw/cmssw/blob/master/RecoHGCal/TICL/plugins/MultiClustersFromTrackstersProducer.cc#L73-L96), for example.

**Review by rovere — CHANGES_REQUESTED** [review]

**Comment by perrotta** [comment]
After this cleaning, is the RecoHGCal/TICL/plugins/MultiClustersFromTrackstersProducer.cc still needed anywhere, or could it be also cleaned up?

It is still included from: 
  - RecoEgamma/EgammaHLTProducers/python/HLTEgPhaseIITestSequence_cff.py

If the intention is to remove its usage from everywhere, perhaps it can also get removed from those test sequences, unless you still intend to have it around for some while.

**Comment by rovere** [comment]
@perrotta thanks for the review and the comments. The `MultiClustersFromTrackstersProducer` is still used and needed to comply with legacy format/reconstruction chains. We are starting to migrate to using our data structure inside `HGCAL` and, eventually, think of migrating everyone else to adopt them (if proved to be convenient, of course). This will be a long term goal/project and is not the purpose of this PR.
I hope this clarifies the situation.

File: RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalTrackster.cc (hunk 2 above)
Comment (lecriste) [reply RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalTrackster.cc]: Right, done.

File: RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalTrackster.cc (hunk 3 above)
Comment (lecriste) [reply RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalTrackster.cc]: Right, done.

File: RecoHGCal/Configuration/python/RecoHGCal_EventContent_cff.py [hunk 4]
```diff
@@ -1,38 +1,18 @@
...
-      'keep *_ticlTrackstersHFNoseMIP_*_*',
-      'keep *_ticlTrackstersHFNoseHAD_*_*',
-      'keep *_ticlTrackstersHFNoseMerge_*_*',
-      'keep *_pfTICL_*_*'
+      trackstersIters +
+      trackstersHFNoseIters +
+      ['keep *_pfTICL_*_*']
       )
     )
-TICL_RECO.outputCommands.extend(TICL_AOD.outputCommands)
```
[the remark this answers is not in the document]
Comment (lecriste) [reply RecoHGCal/Configuration/python/RecoHGCal_EventContent_cff.py:35]: OK, done.

File: Validation/HGCalValidation/interface/HGVHistoProducerAlgo.h [hunk 5]
```diff
@@ -281,31 +282,32 @@ class HGVHistoProducerAlgo {
...
                                    const hgcal::RecoToSimCollection& recSimColl,
                                    const hgcal::SimToRecoCollection& simRecColl) const;
-  void fill_simcluster_histos(const Histograms& histograms,
-                              std::vector<SimCluster> const& simclusters,
+  void fill_simCluster_histos(const Histograms& histograms,
+                              std::vector<SimCluster> const& simClusters,
                               unsigned int layers,
                               std::vector<int> thicknesses) const;
-  void fill_simclusterassosiation_histos(const Histograms& histograms,
+  void fill_simClusterassosiation_histos(const Histograms& histograms,
```
[the remark this answers is not in the document]
Comment (lecriste) [reply Validation/HGCalValidation/interface/HGVHistoProducerAlgo.h]: Yes, done.

File: RecoHGCal/Configuration/python/RecoHGCal_EventContent_cff.py [hunk 6]
```diff
@@ -1,38 +1,18 @@
...
-      'keep *_ticlMultiClustersFromTrackstersEM_*_*',
-      'keep *_ticlMultiClustersFromTrackstersHAD_*_*',
-      'keep *_ticlMultiClustersFromTrackstersTrk_*_*',
-      'keep *_ticlMultiClustersFromTrackstersTrkEM_*_*',
-      'keep *_ticlMultiClustersFromTrackstersMIP_*_*',
-      'keep *_ticlMultiClustersFromTrackstersMerge_*_*',
-      'keep *_ticlMultiClustersFromSimTracksters_*_*',
-      )
-    )
+trackstersIters = ['keep *_ticlTracksters'+iteration+'_*_*' for iteration in ticlIterLabelsMerge]
```
[the remark this answers is not in the document]
Comment (lecriste) [reply RecoHGCal/Configuration/python/RecoHGCal_EventContent_cff.py:5]: Yes, there are no `ticlTrackstersMIP` produced at the moment.
The new code will automatically include them when available.

File: RecoHGCal/Configuration/python/RecoHGCal_EventContent_cff.py [hunk 7]
```diff
@@ -1,38 +1,18 @@
...
+trackstersHFNoseIters = ['keep *_ticlTrackstersHFNose'+iteration+'_*_*' for iteration in ticlIterLabelsMerge]
```
[the remark this answers is not in the document]
Comment (lecriste) [reply RecoHGCal/Configuration/python/RecoHGCal_EventContent_cff.py]: Actually we prefer to keep the HFNose event content independent.
The last commit reverts this change.

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
For a quick check of the effect of the removal of the HGCal multiclusters from the reco output I run some 100 events of wf 23234 (phase2 geometry D49, no pileup). 

- The HGCal collections not present in output any more are the following (accounting for  the 1.8% of the overall reco content in this workflow):
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
    725.0 ->         0.0       -725     OLDO  -0.06     recoHGCalMultiClusters_ticlMultiClustersFromTrackstersHAD__RECO.
  11726.1 ->         0.0     -11726     OLDO  -0.89     recoHGCalMultiClusters_ticlMultiClustersFromTrackstersMerge__RECO.
   2976.5 ->         0.0      -2977     OLDO  -0.23     recoHGCalMultiClusters_ticlMultiClustersFromTrackstersEM__RECO.
   1558.1 ->         0.0      -1558     OLDO  -0.12     recoHGCalMultiClusters_ticlMultiClustersFromTrackstersTrkEM__RECO.
   6790.8 ->         0.0      -6791     OLDO  -0.52     recoHGCalMultiClusters_ticlMultiClustersFromTrackstersTrk__RECO.
-------------------------------------------------------------
  1311065 ->     1287233     -23831            -1.8     ALL BRANCHES
```


- The CPU timing saved by the corresponding modules which are not run any more is indeed rather limited, order of a ms in this (no PU) wf:
```
The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
     removed      -0.01%         0.28 ms/ev ->         0.00 ms/ev ticlMultiClustersFromTrackstersTrk
     removed      -0.01%         0.22 ms/ev ->         0.00 ms/ev ticlMultiClustersFromTrackstersTrkEM
     removed      -0.01%         0.23 ms/ev ->         0.00 ms/ev ticlMultiClustersFromTrackstersHAD
     removed      -0.01%         0.41 ms/ev ->         0.00 ms/ev ticlMultiClustersFromTrackstersMerge
     removed      -0.01%         0.23 ms/ev ->         0.00 ms/ev ticlMultiClustersFromTrackstersEM
  ---------- ------------     --------                  ----       ------------
Job total:  3.31021 s/ev ==> 3.26741 s/ev
```

**Comment by perrotta** [comment]
+1
- This PR replaces the usage of HGCal MultiCluster with Trackster, thus cleaning the HGCal reconstruction step from producing the now deprecate multicluster's
- The only differences in the output of the Jenkins tests just witness the removal of the multiclusters from the reco output, with no changes anywhere else, as expected

File: Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py (hunk 1 above)
Comment (lecriste) [reply Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py:23]: Yes, I will take care of this in the next PR.

**Comment by lecriste** [comment]
@cms-sw/upgrade-l2 can you please review this PR?

**Comment by srimanob** [comment]
Hi @lecriste 
Just to confirm, with this PR, there is no need for analysis using AOD for trackster collection? Bring in/out of AOD collections does not seem a good plan. Thanks.

**Comment by lecriste** [comment]
Hi @srimanob, since we don't know about analysis consuming AOD nowadays for HGCAL purposes, we can add the requested collections in another PR when asked for (as Marco [said](https://github.com/cms-sw/cmssw/pull/33488#issuecomment-824171600)).

**Comment by srimanob** [comment]
+Upgrade

This PR may need the follow-up PR later for the content of AOD (for analysis).

**Comment by silviodonato** [comment]
>  No changes are expected, apart from naming shift ("MultiCluster" -> "Trackster") in DQM output.

@lecriste I've made a quick check on https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-04-23-1100+22df90/42386/dqm-histo-comparison-summary.html
and I see that indeed 978 are added and 978 are removed.
Many plots are just renamed, as expected, but some plots show some changes.
Eg. in  23234.0  HGCAL / HGCalValidator / ticlMultiClustersFromSimTracksters  (ticlSimTracksters)
https://tinyurl.com/yk4hafwa
vs
https://tinyurl.com/yecxom8s

image
image

Is it expected?

**Comment by lecriste** [comment]
Thanks for the check @silviodonato.
Those changes are not intended and arose from a superfluous loop introduced while addressing a review comment.
I have fixed it in the latest commit.

**Comment by silviodonato** [comment]
please test

**Comment by fwyzard** [comment]
let's see if this works...

**Comment by fwyzard** [comment]
assign hgcal-dpg

**Comment by fwyzard** [comment]
mhm... looks like it doesn't :(

**Comment by qliphy** [comment]
@cms-sw/hgcal-dpg-l2

**Comment by rovere** [comment]
> @cms-sw/hgcal-dpg-l2

what does this command do, explicitly?

**Comment by qliphy** [comment]
> > @cms-sw/hgcal-dpg-l2
> 
> what does this command do, explicitly?

This currently gives the possibility to tag experts from DPGs/POGs in a chat.
https://github.com/cms-sw/cms-bot/pull/1532

**Comment by fwyzard** [comment]
ok, but one still cannot request the DPGs to "sign" a PR ?

**Comment by srimanob** [comment]
assign @cms-sw/hgcal-dpg-l2

**Comment by rovere** [comment]
assign hgcal-dpg

**Comment by qliphy** [comment]
@hgcal-dpg

**Comment by qliphy** [comment]
assign hgcal-dpg

**Comment by fwyzard** [comment]
I don't see any labels for the DPGs on GitHub.

**Comment by qliphy** [comment]
> I don't see any labels for the DPGs on GitHub.

Right, probably some more should be done on "CMSSW_CATEGORIES":
https://github.com/cms-sw/cms-bot/blob/master/categories_map.py  
in order to be picked up by "assign"
https://github.com/cms-sw/cms-bot/blob/aadcdcbfd5bc4b4ef46076a0b39f0072993746da/process_pr.py#L163-L172

@smuzaffar @silviodonato  may comment more, or we can discuss at today's ORP.

**Comment by rovere** [comment]
Thanks a lot @qliphy for following up further on this!

**Comment by smuzaffar** [comment]
I am looking in to the label issue

**Comment by smuzaffar** [comment]
bot was not allowing `-` in the category name. This is fixed now and category assignment and labels should be there in few minutes

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
> bot was not allowing `-` in the category name. This is fixed now and category assignment and labels should be there in few minutes

So, what is the correct syntax to use? 
`assign hgcal-dpg`?

**Comment by smuzaffar** [comment]
yes "assign category" is for asking L2's of `category` to sign the PR. `@cms-sw/category-l2` is to ping the members of `category`. See http://cms-sw.github.io/cms-bot-cmssw-cmds.html

**Comment by perrotta** [comment]
> * Reco comparison results: 3134 differences found in the comparisons

New differences are in the DD4Hep workflow 11634.911, a recurring issue independent from this PR...

**Comment by jfernan2** [comment]
> Those changes are not intended and arose from a superfluous loop introduced while addressing a review comment.
> I have fixed it in the latest commit.

It looks like the differences are still there sorry:
PR: https://tinyurl.com/yzjwldv6
Base: https://tinyurl.com/yzmq8rjk

File: Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc [hunk 8]
```diff
@@ -2504,94 +2497,99 @@ void HGVHistoProducerAlgo::multiClusters_to_CaloParticles(const Histograms& hist
...
+        cpPair.second = 1.;
+        LogDebug("HGCalValidator") << "Trackster Id: \t" << tstId << "\t CP id: \t" << cpPair.first << "\t score \t"
+                                   << cpPair.second << std::endl;
+        histograms.h_score_trackster2caloparticle[count]->Fill(cpPair.second);
       }
+      continue;
+    }
+
+    for (const auto lcId : tracksters[tstId].vertices()) {
+      const auto& hits_and_fractions = layerClusters[lcId].hitsAndFractions();
```
Comment (rovere) [inline Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc:2584]: I believe you are missing the correct normalization using the `vertex_multiplicity`, right?
LayerClusters are built **before** `Tracksters` and have no concept, so far, of energy sharing. This is added on top while building Tracksters and were propagated correctly while creating `MultiClustersFromTracksters`. We should do the same here.

**Review by rovere — CHANGES_REQUESTED** [review]

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
Hi @lecriste 
There seems to be some difference in the Shared Energy per Calo Particle in each layer, at least:
PR: https://tinyurl.com/ygzfqyww
BASE: https://tinyurl.com/yzbsjfsl
I do not know if this is tolerated or not, other differences may exist also, I have not compared all MEs/folders

**Comment by lecriste** [comment]
Hi @jfernan2, those differences are due to numeric precision.
On 10 events of that workflow (~23 multiClusters/tracksters per event) I found only these 3 differences:
```
< obj: 2	energy: 1.63313
> obj: 2	energy: 1.63312

< obj: 3	energy: 99.2324
> obj: 3	energy: 99.2325

< obj: 3	energy: 5.20563
> obj: 3	energy: 5.20564
```

**Comment by jfernan2** [comment]
OK, but mean and stddev of that (e.g.) ME is changing at the 10% level, while the changes you quote are at 0.01% level, is this really expected?

**Comment by lecriste** [comment]
Yes, the differences I quoted enter as denominator in the fraction used to fill those MEs.

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
> Yes, the differences I quoted enter as denominator in the fraction used to fill those MEs.

Just to understand: if the ratio has a difference otself if order 10%, while the denominator has a difference of the order of 0,01%, shouldn't you explain why the numerator (what else?) has this 10% difference?

**Comment by jfernan2** [comment]
> Yes, the differences I quoted enter as denominator in the fraction used to fill those MEs.

In any case, @lecriste you should describe this change in the PR description for the records, since in principle the PR implied a folder renaming only. Thanks

**Comment by lecriste** [comment]
@perrotta the numerator is unchnaged. The difference in the fraction (of the same order of the one in the denominator) produces an overflow migration resulting in a different Mean and StdDev of the affected MEs (only the last bin differs between [the two MEs](https://github.com/cms-sw/cmssw/pull/33488#issuecomment-831185312)). As I wrote in the PR description, the unbinned histogram mean is identical at the numerical precision level.

**Comment by perrotta** [comment]
+1
- No reco related updates since previous reco signature in https://github.com/cms-sw/cmssw/pull/33488#issuecomment-826591384
- This PR replaces the usage of HGCal MultiCluster with Trackster, thus cleaning the HGCal reconstruction step from producing the now deprecate multicluster's
The only differences in the event output of the Jenkins tests just witness the removal of the multiclusters from the reco output, with no changes anywhere else, as expected; changes observed in a few DQM plots are at the level of the numerical precision

**Comment by lecriste** [comment]
@cms-sw/upgrade-l2 can you please review this PR?

**Comment by perrotta** [comment]
> @cms-sw/upgrade-l2 can you please review this PR?

@hgcal-dpg since you requested to have this PR also assigned to you please don't forget to sign (if you think so, of course)

**Comment by rovere** [comment]
@perrotta thanks for the kind reminder. In fact, I'm not yet fully convinced that this PR will introduce no regression. A separate discussion is ongoing with the developer/proponent.
As soon as I'm confident, I'll approve the PR.

**Comment by rovere** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

This PR is to review the eventcontent for HGCal and to update names used in validation sequence. No change is expected.

**Comment by qliphy** [comment]
+1

**Pull request merged**